exercises/takudzwa_makoni/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,163 @@
+};
+
+
+//--------------------------------------------------------------------------
+
+//------------------------- shape subclasses -------------------------------
+
+class square : public shape {
+
+private:
+
+	double sideLength;
+	string name;
+	int numOfSides;
+
+public:
+
+	virtual void getNumOfSides() { cout << numOfSides << endl; }
+	virtual void getName() { cout << name << endl; }
+
+	virtual void getArea() { cout << sideLength * sideLength << endl; }
+	virtual void getPerimeter() { cout << 4 * sideLength << endl; }
+
+	//contructor
+	square::square() {
+		sideLength = 1;
+		name = "square";
+		numOfSides = 4;
+	};
+};
+
+class rectangle : public shape {
+
+private:
+
+	double sideWidth, sideHeight;
+	string name;
+	int numOfSides;
+
+public:
+
+	virtual void getNumOfSides() { cout << numOfSides << endl; }
+	virtual void getName() { cout << name << endl; }
+
+	virtual void getArea()		{ cout << sideHeight * sideWidth << endl; }
+	virtual void getPerimeter()	{ cout << (2 * sideWidth) + (2 * sideHeight) << endl; }
+
+	rectangle::rectangle() {
+		sideWidth = 1;
+		sideHeight = 2;
+		name = "rectangle";
+		numOfSides = 4;
+	};
+};
+
+class isoTriangle : public shape { // isosceles triangle class
+
+private:
+
+	double base, height;
+	string name;
+	int numOfSides;
+
+
+public:
+
+	virtual void getNumOfSides()	{ cout << numOfSides << endl; }
+	virtual void getName()			{ cout << name << endl; }
+
+	virtual void getArea()		{ cout << 0.5 * base* height << endl; }
+	virtual void getPerimeter() { cout << 2 * (sqrt(height * height) + ((base * base) / 4)) << endl; }
+
+	//constructor
+	isoTriangle::isoTriangle() {
+		base	= 1;
+		height	= 2;
+		name = "isosceles-triangle";
+		numOfSides = 3;
+	};
+};
+
+
+class circle : public shape { // circle class
+
+private:
+
+	double radius;
+	string name;
+	int numOfSides;
+
+public:
+
+	virtual void getNumOfSides() { cout << numOfSides << endl; }
+	virtual void getName() { cout << name << endl; }
+
+	virtual void getArea() { cout << pi * radius * radius << endl; }
+	virtual void getPerimeter() { cout << pi * 2 * radius << endl; }
+
+	//constructor
+	circle::circle() {
+		radius = 1;
+		name = "circle";
+		numOfSides = 1;
+	};
+
+};
+
+//--------------------------------------------------------------
+
+//------------------- function definitions ---------------------
+
+
+void printMatchForType(vector<shape>& shapes) {};		// print match for type. Wont mutate vector so will pass by reference.
+void printMatchForSide(vector<shape>& shapes) {};		// print match for number of side
+void printOrderForVol(vector<shape>& shapes) {};		// print order for volume (descending)
+void printorderForPerim(vector<shape>& shapes) {};		// print order for perimeter (descending)
+
+
+//------------------------- main -------------------------------
+
+int main() {
+
+
+	square x;
+	circle y;
+	rectangle z;
+	isoTriangle p;
+
+	vector<shape*> shapeVector({ &x, &y, &z, &p });
```

Comment:
Interesting approach - technically not wrong but you should ensure you are familiar with `smart pointers` (I think Scott Meyers has a chapter specifically on the `vector` problem you encounter here) so that you can code correctly where scope changes mean that ownership transfer is required.